DSD-338: Chakra template components

EdwinGuzman opened the pull request:
Fixes JIRA ticket [DSD-338](https://jira.nypl.org/browse/DSD-338)

## This PR does the following:
- Creates the following components based on the previous DS layout CSS rules: `Template`, `TemplateHeader`, `TemplateContentBreakout`, `TemplateContent`, `TemplateContentTop`, `TemplateContentPrimary`, `TemplateContentSidebar`, and `TemplateFooter`.
- The components above must be rendered in the right DOM structure for the desired layout. This is the children component approach.
- This PR also creates a `TemplateAppContainer` component where each section can be passed as a prop. This is the prop-based approach. Through this approach, only one component is needed but each section must be passed in the appropriate prop name.

## How has this been tested?

Locally and storybook.

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have updated the Storybook documentation accordingly.
- [ ] All new and existing tests passed.

### Front End Review:

<!--- This step is done AFTER creating a PR. -->
<!--- Tugboat creates a static Storybook preview URL once the PR is created. -->
<!--- Copy the URL to the relevant Storybook page here. -->

- [ ] View [the example in Storybook]()

EdwinGuzman committed:
Adding template components and TemplateAppContainer component composed from Chakra components

EdwinGuzman commented:
After working on this, I do think that the prop-based approach is cleaner than the children component-based approach. For the following, assume each variable passed is a DOM element.

## prop-based:
```
<TemplateAppContainer
  header={header}
  contentBreakout={contentBreakout}
  sidebar={sidebar}
  contentTop={contentTop}
  contentSidebar={contentSidebar}
  contentPrimary={contentPrimary}
  footer={footer}
/>
```

## children component-based:
```
<Template>
  <TemplateHeader>
    {header}
    <TemplateContentBreakout>{contentBreakout}</TemplateContentBreakout>
  </TemplateHeader>
  <TemplateContent sidebar={sidebar}>
    <TemplateContentTop>{contentTop}</TemplateContentTop>
    <TemplateContentSidebar>{contentSidebar}</TemplateContentSidebar>
    <TemplateContentPrimary>{contentPrimary}</TemplateContentPrimary>
  </TemplateContent>
  <TemplateFooter>{footer}</TemplateFooter>
</Template>
```

EdwinGuzman commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr687-ejsadey7iwfitkijlrdkgvdswrwsyoh8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/614a6138d7000a00be4788e1

EdwinGuzman committed:
Export the template components

EdwinGuzman commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr687-ejsadey7iwfitkijlrdkgvdswrwsyoh8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/614a6138d7000a00be4788e1

EdwinGuzman committed:
Merging and fixing conflicts from development

EdwinGuzman commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr687-ejsadey7iwfitkijlrdkgvdswrwsyoh8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/614a6138d7000a00be4788e1

bigfishdesign13 commented on src/components/Template/Template.stories.mdx:
```diff
@@ -0,0 +1,410 @@
+render a consistent mobile and desktop layout. More information on how they
+should be composed below.
+
+Basic component structure:
```
If we are going to suggest using `TemplateAppContainer` as the best practice, we should move that to the top of the story.

EdwinGuzman replied:
We haven't decided on this yet and now is the time to do so. Should we suggest the `TemplateAppContainer` or the individual components?

I'm okay with recommending `TemplateAppContainer` but it would update a lot of this PR, which I'm okay with doing.

bigfishdesign13 commented on src/components/Template/Template.stories.mdx:
```diff
@@ -0,0 +1,410 @@
+In order to render a sidebar, two steps must be taken.
+
+1. The `TemplateContent` component needs to be aware of it. Pass the location
+   of the sidebar to the `sidebar` prop as either "left" or "right".
```
Later in the story, you state that the `TemplateContentSidebar` child component needs to be properly situated in the code structure to determine its left or right side placement.  What happens if `TemplateContentSidebar` come before `TemplateContentPrimary` in the code, but the `sidebar` prop is set to `right`?

EdwinGuzman replied:
The `TemplateContentSidebar` component won't render properly.

bigfishdesign13 replied:
In theory, we should be able to control the positioning of the sidebar with CSS.  Which would mean that the ordering of the component in the code should not matter.

bigfishdesign13 replied:
I took a look at how screen readers deal with the flex order attribute.  As you noted, they don't care about that value and simply look at the order in the DOM.  So yes, the DOM order will reign supreme.

bigfishdesign13 commented on src/components/Template/Template.stories.mdx:
```diff
@@ -0,0 +1,410 @@
+
+<Description of={TemplateHeader} />
+
+### TemplateContentBreakout
```
The name of this component is a bit confusing.  It sounds like it should be inside `TemplateContent`.  `TemplateBreakout` would make the component more agnostic and not appear tied to `TemplateContent`.

EdwinGuzman replied:
That's true. In the previous examples, this was only used in the header. So maybe it should be `TemplateHeaderBreakout` since only components there go edge-to-edge.

bigfishdesign13 replied:
Yes, that works.  The breakout mixin is used in other components, so that is something that should be moved into a global file at some point.

bigfishdesign13 commented on src/components/Template/Template.tsx:
```diff
@@ -0,0 +1,225 @@
+ * This is meant to render its children from edge to edge and is most useful
+ * for the `Breadcrumbs` and `Hero` components, and other banner-like components.
+ */
+const TemplateContentBreakout = (
```
As I mentioned in the story file, I think the name of this child component should be changed to `TemplateBreakout`, which would bring it into alignment with the prop name in `TemplateAppContainer`.

bigfishdesign13 commented on src/components/Template/Template.tsx:
```diff
@@ -0,0 +1,225 @@
+};
+
+/**
+ * This is most useful component to render content on the page. This renders an
```
`This component is most useful for rendering content on the page.`

bigfishdesign13 commented on src/components/Template/Template.tsx:
```diff
@@ -0,0 +1,225 @@
+/**
+ * This is most useful component to render content on the page. This renders an
+ * HTML `<main>` element. This takes a `sidebar` prop with optional "left" or
+ * "right" values. This will pass down the correct Chakra styling for the
```
Does this need to reference `Chakra`?  Can't it just state that the correct styling will be applied?

EdwinGuzman replied:
Yeah, at a high level that's fine to say. I just figured it'd be easier to debug if it was stated that it is styled through Chakra.

bigfishdesign13 commented on src/components/Template/Template.tsx:
```diff
@@ -0,0 +1,225 @@
+/**
+ * This component is used to render content in a column when there must be a
+ * sidebar component on either its left or right side. It must go inside the
+ * `TemplateContent` component. An optional `sidebar` prop value of "left" or
```
Using the `sidebar` prop here seems like overkill.  Why not implment the prop only at the parent level?

EdwinGuzman replied:
For `TemplateContentPrimary` and `TemplateContentSidebar`, this is implemented as a prop so that the parent can add the value at its own level. So while the user doesn't need to set it when they render it, the parent `TemplateContent` internally passes this value. So yes, this prop value is set at the parent level right now, but it still needs to be declared.

bigfishdesign13 commented on src/components/Template/Template.tsx:
```diff
@@ -0,0 +1,225 @@
+ * `TemplateContentPrimary` component. If this is a left sidebar, it needs to be
+ * rendered before the `TemplateContentPrimary` component. If this is a right
+ * sidebar, it needs to be rendered after the `TemplateContentPrimary` component.
+ * An optional `sidebar` prop value of "left" or "right" can be passed to render
```
Using the `sidebar` prop here seems like overkill.  Why not implment the prop only at the parent level?

bigfishdesign13 commented on src/components/Template/Template.tsx:
```diff
@@ -0,0 +1,225 @@
+
+/**
+ * This optional component should be the last child of the `Template`
+ * component. This is rendered as an HTML `<footer>` element.
```
This should explain that this UI element spans the full width of the screen.

bigfishdesign13 commented on src/theme/components/template.ts:
```diff
@@ -0,0 +1,115 @@
+  marginRight: "calc(-50vw + 50%)",
+};
+// From the `wrapper` SCSS mixin.
+// TODO: Eventually we may be able to put shared JS style objects
```
Yes!

bigfishdesign13 commented on src/theme/components/template.ts:
```diff
@@ -0,0 +1,115 @@
+    boxSizing: "border-box",
+    color: "ui.black",
+    fontFamily: "body",
+    fontSize: "16px",
```
If you are including these next few styles here, does this mean our base text styles are not included anywhere in the theme yet?

EdwinGuzman replied:
The family was but not the size. I updated it.

bigfishdesign13 replied:
Can the default font size be included in the base theme styles?

bigfishdesign13 commented on src/theme/components/template.ts:
```diff
@@ -0,0 +1,115 @@
+    marginBottom: 8, // --space-l
+  },
+};
+const TemplateContentBreakout = {
```
In alignment with my other comments, this should be changed to `TemplateBreakout`.

bigfishdesign13 commented on src/theme/components/template.ts:
```diff
@@ -0,0 +1,115 @@
+  baseStyle: {
+    flex: "0 0 25%",
+    order: { md: "1" },
+    width: "225px",
```
Remove this.

EdwinGuzman replied:
Okay. The description in the JIRA ticket said the sidebar should have this fixed value so that's why it's here.

bigfishdesign13 commented on src/theme/components/template.ts:
```diff
@@ -0,0 +1,115 @@
+};
+const TemplateContentSidebar = {
+  baseStyle: {
+    flex: "0 0 25%",
```
We want the sidebar to have a fixed width, so set this to `flex: "0 0 225px`.

bigfishdesign13 commented on src/theme/components/template.ts:
```diff
@@ -0,0 +1,115 @@
+  },
+  variants: {
+    left: {
+      marginRight: 12, // --space-xl
```
To be responsive, change to:
```marginRight: { base: 0, md: 12 }```

bigfishdesign13 commented on src/theme/components/template.ts:
```diff
@@ -0,0 +1,115 @@
+  variants: {
+    left: {
+      marginRight: 12, // --space-xl
+      flexBasis: { md: "calc(25% - calc(4 * 16px))" },
```
With the updates to `baseStyle`,  this can be removed.

bigfishdesign13 commented on src/theme/components/template.ts:
```diff
@@ -0,0 +1,115 @@
+      flexBasis: { md: "calc(25% - calc(4 * 16px))" },
+    },
+    right: {
+      marginLeft: 12, // --space-xl
```
To be responsive, change to:
```marginLeft: { base: 0, md: 12 }```

bigfishdesign13 commented on src/theme/components/template.ts:
```diff
@@ -0,0 +1,115 @@
+    },
+    right: {
+      marginLeft: 12, // --space-xl
+      flexBasis: { md: "calc(25% - calc(4 * 16px))" },
```
With the updates to `baseStyle`, this can be removed.

bigfishdesign13 reviewed (changes requested):
There's a lot going on here.  I love it.  I have a lot of questions in my comments, as well as some necessary style changes.

EdwinGuzman committed:
Some updates based on feedback

EdwinGuzman commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr687-ejsadey7iwfitkijlrdkgvdswrwsyoh8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/614a6138d7000a00be4788e1

EdwinGuzman committed:
Merging from chakra-phase2 and fixing conflicts

EdwinGuzman committed:
Updating name for the template breakout component

EdwinGuzman committed:
Merge branch 'chakra-phase2' into DSD-338/chakra-template

* chakra-phase2:
  Updating error CSS variable name in previously missed files
  Updating the NYPL colors and names

EdwinGuzman committed:
Updating the Template Storybook docs

EdwinGuzman commented:
Updated:
- `TemplateAppContainer` is now at the top of the Storybook documentation and is the recommended approach.
- Added a more detail example for the template children component docs.

EdwinGuzman committed:
Merging from chakra-phase2 and fixing conflicts

EdwinGuzman committed:
Updating template docs and adding chakra-ui/system

EdwinGuzman committed:
Merge branch 'chakra-phase2' into DSD-338/chakra-template

* chakra-phase2:
  Adding chakra-ui/system

EdwinGuzman commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr687-ejsadey7iwfitkijlrdkgvdswrwsyoh8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/614a6138d7000a00be4788e1

bigfishdesign13 reviewed (approved):
The updates look good.
